Swiftfin/Views/LibraryView.swift

```diff
@@ -21,87 +22,75 @@ struct LibraryView: View {
 	var defaultFilters = LibraryFilters(filters: [], sortOrder: [.ascending], withGenres: [], tags: [], sortBy: [.name])
 
 	@State
-	private var tracks: [GridItem] = Array(repeating: .init(.flexible()), count: Int(UIScreen.main.bounds.size.width) / 125)
+	private var tracks: [GridItem] = Array(repeating: .init(.flexible(), alignment: .top),
+	                                       count: Int(UIScreen.main.bounds.size.width) / 125)
 
 	func recalcTracks() {
-		tracks = Array(repeating: .init(.flexible()), count: Int(UIScreen.main.bounds.size.width) / 125)
+		tracks = Array(repeating: .init(.flexible(), alignment: .top), count: Int(UIScreen.main.bounds.size.width) / 125)
 	}
 
-	var body: some View {
-		Group {
-			if viewModel.isLoading == true {
-				ProgressView()
-			} else if !viewModel.items.isEmpty {
-				VStack {
-					ScrollView(.vertical) {
-						Spacer().frame(height: 16)
-						LazyVGrid(columns: tracks) {
-							ForEach(viewModel.items, id: \.id) { item in
-								if item.type != "Folder" {
-									Button {
-										libraryRouter.route(to: \.item, item)
-									} label: {
-										PortraitItemView(item: item)
-									}
-								}
-							}
-						}.onRotate { _ in
-							recalcTracks()
-						}
-						if viewModel.hasNextPage || viewModel.hasPreviousPage {
-							HStack {
-								Spacer()
-								HStack {
-									Button {
-										viewModel.requestPreviousPage()
-									} label: {
-										Image(systemName: "chevron.left")
-											.font(.system(size: 25))
-									}.disabled(!viewModel.hasPreviousPage)
-									Text(L10n.pageOfWithNumbers(String(viewModel.currentPage + 1), String(viewModel.totalPages)))
-										.font(.subheadline)
-										.fontWeight(.medium)
-									Button {
-										viewModel.requestNextPage()
-									} label: {
-										Image(systemName: "chevron.right")
-											.font(.system(size: 25))
-									}.disabled(!viewModel.hasNextPage)
-								}
-								Spacer()
+	@ViewBuilder
+	private var loadingView: some View {
+		ProgressView()
+	}
+
+	@ViewBuilder
+	private var noResultsView: some View {
+		L10n.noResults.text
+	}
+
+	@ViewBuilder
+	private var libraryItemsView: some View {
+		DetectBottomScrollView {
```

Comment:
As you wrote, it seems like a good way to create and use `DetectBottomScrollView`.

But wouldn't it be simpler to use `onAppear` for each cell and compare it with the last item of `viewModel.items` rather than that `DetectBottomScrollView`, and there will be less side effects?

```swift
PortraitItemButton(item: item) { item in
	libraryRouter.route(to: \.item, item)
}
.onAppear {
	  guard viewModel.hasNextPage,
	  	item == viewModel.items.last else { return }
	  viewModel.request~
}
```

like this

Reply:
I actually did try that, however `ScrollView`'s lifecycle prevents this from acting as expected. It will render items twice in quick succession (once when added to the `ScrollView` and then again once the size of the `ScrollView` is calculated). Because of this, I experienced errors with the page calculations as `onAppear()` was run multiple times unexpectedly. 

`DetectBottomScrollView` is actually a lot better than `onAppear(...)` in this case with not side effects, what side effects are you pointing to?

Reply:
If we use `onAppear` for cells in the current structure, doesn't it follow the lifecycle of `LazyVGrid`?
So I thought it would work normally.

Anyway, I don't know because I haven't used that code myself.
The overlapping use of A often had side effects in certain situations lol.

Once it works normally in the current situation, it will be fine.